verifier/src/main/java/com/aws/jverify/verifier/compiler/ContractCompiler.java

```diff
@@ -0,0 +1,190 @@
+package com.aws.jverify.verifier.compiler;
+
+import com.aws.jverify.common.Common;
+import com.aws.jverify.generated.*;
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.tree.TreeInfo;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+Extracts contracts from constructor, method or loop bodies
+ */
+public class ContractCompiler {
+    public final JavaToDafnyCompiler compiler;
+
+    public ContractCompiler(JavaToDafnyCompiler compiler) {
+        this.compiler = compiler;
+    }
+
+    public List<JCTree.JCStatement> translateHeader(JCTree.JCStatement statement, 
+                                                    MethodOrLoopContract contract,
+                                                    boolean allowFooterButNotHeader,
+                                                    boolean reportErrors) {
+        var statements = statement instanceof JCTree.JCBlock block
+                ? block.getStatements()
+                : List.of(statement);
+        return translateHeader(statements, contract, allowFooterButNotHeader, reportErrors);
+    }
+    
+    public List<JCTree.JCStatement> translateHeader(List<JCTree.JCStatement> statements,
+                                                    MethodOrLoopContract contract,
+                                                    boolean allowFooter,
+                                                    boolean reportErrors) {
+
+        JCTree.JCStatement superOrThis = null;
+        var first = statements.isEmpty() ? null : statements.getFirst();
+        if ((first instanceof JCTree.JCExpressionStatement expressionStatement
+                && expressionStatement.getExpression() instanceof JCTree.JCMethodInvocation invocation)) {
+            var isSuperOrThisCall = invocation.getMethodSelect() instanceof JCTree.JCIdent ident && 
+                    (ident.name == ident.name.table.names._super || ident.name == ident.name.table.names._this);
+            if (isSuperOrThisCall) {
+                superOrThis = first;
+            }
```

Comment:
nit: There's a `TreeInfo` utility method that does the same thing -
```suggestion
            if (TreeInfo.isConstructorCall(invocation)) {
                superOrThis = first;
            }
```

Reply:
I don't see `isConstructorCall`. I see `isConstructor` but that's not what I need. There's `isSuperCall` which is good but I don't see `isThisCall`